pzsd_bot/cogs/reminders/reminders.py

```diff
@@ -0,0 +1,273 @@
+        self.scheduler.cancel_all()
+
+    async def reschedule_reminder(self, reminder_data: Row) -> None:
+        logger.info("Rescheduling recurring reminder (id=%s)", reminder_data.id)
+
+        self.scheduler.schedule(
+            run_at=reminder_data.remind_at,
+            task_id=f"reminder_{reminder_data.id}",
+            coroutine=self.send_reminder(reminder_data),
+        )
+
+    async def send_reminder(self, reminder_data: Row) -> None:
+        logger.info("Attempting to send reminder (id=%s)", reminder_data.id)
+
+        channel = self.bot.get_channel(reminder_data.channel_id)
+        if channel is None:
+            # channel may not be cached yet
+            logger.debug("Couldn't find channel in cache, attempting to fetch")
+            try:
+                channel = await self.bot.fetch_channel(reminder_data.channel_id)
+            except Exception as e:
+                logger.warning(
+                    "Failed to send reminder (id=%s), can't find channel (id=%s) with error: %s",
+                    reminder_data.id,
+                    reminder_data.channel_id,
+                    e,
+                )
+                async with Session.begin() as session:
+                    await session.execute(
+                        update(reminder)
+                        .values(status=ReminderStatus.failed)
+                        .where(reminder.c.id == reminder_data.id)
+                    )
+
+                return
+
+        original_message = channel.get_partial_message(
+            reminder_data.original_message_id
+        )
+        embed = Embed(
+            description=reminder_data.reminder_text,
+            colour=Colour.blurple(),
+        )
+        await original_message.reply("Here's your reminder:", embed=embed)
+
+        logger.info("Reminder sent (id=%s)", reminder_data.id)
+
+        if reminder_data.is_recurring:
+            new_remind_at = reminder_data.remind_at + pendulum.duration(
+                seconds=reminder_data.recurrence_interval
+            )
+            async with Session.begin() as session:
+                result = await session.execute(
+                    update(reminder)
+                    .values(remind_at=new_remind_at)
+                    .where(reminder.c.id == reminder_data.id)
+                    .returning(reminder)
+                )
+                reminder_data = result.one()
+
+            asyncio.create_task(self.reschedule_reminder(reminder_data))
+        else:
+            logger.debug("Deleting reminder with id=%s", reminder_data.id)
+            async with Session.begin() as session:
+                await session.execute(
+                    delete(reminder).where(reminder.c.id == reminder_data.id)
+                )
+
+    @Cog.listener()
+    async def on_message(self, message: Message) -> None:
+        if message.author == self.bot.user:
+            return
+
+        if m := REMINDER_PATTERN.search(message.content):
```

Comment:
consider doing an early return instead of nesting the entire rest of the function under this if statement because it ends up going pretty deep